### PR TITLE
ethapi: GetTransactionReceipt panic fix

### PR DESCRIPTION
**Pull request by rhaps107**
Panic occurs when core.GetReceipt may return nil receipt.

#15408
#14432

### CODE REVIEW

**Comment by GitCop** [comment]
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: a66118744ff057f254e24d880cab58b5fe10e9b9
  * Commits must be prefixed with the package(s) they modify

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

File: internal/ethapi/api.go [hunk 1]
```diff
@@ -1006,6 +1006,9 @@ func (s *PublicTransactionPoolAPI) GetTransactionReceipt(hash common.Hash) (map[
 		return nil, nil
 	}
 	receipt, _, _, _ := core.GetReceipt(s.b.ChainDb(), hash) // Old receipts don't have the lookup data available
+	if receipt == nil {
+		return nil, nil
+	}
```
Comment (holiman) [inline internal/ethapi/api.go]: Looks good to me in general, but I would suggest that you signal an error here, something like: 
```
     return nil, fmt.Errorf("Missing receipt for hash %v", hash)
```

Comment (rhaps107) [reply internal/ethapi/api.go]: done.
for nil transaction response too.

**Comment by holiman** [comment]
Thanks!

I was a bit hasty in my example code.. Now it shows:
```
> eth.getTransactionReceipt("0xf2e3cd05eac4934a8271924668d8d540edc4313e8f0200f9d51f7f86ea0dd0dc")
Error: Missing transaction for hash [242 227 205 5 234 196 147 74 130 113 146 70 104 216 213 64 237 196 49 62 143 2 0 249 213 31 127 134 234 13 208 220]
    at web3.js:3143:20
    at web3.js:6347:15
    at web3.js:5081:36
    at <anonymous>:1:1

```
With my fix: 
```
> eth.getTransactionReceipt("0xf2e3cd05eac4934a8271924668d8d540edc4313e8f0200f9d51f7f86ea0dd0dc")
Error: Missing transaction for hash 0xf2e3cd05eac4934a8271924668d8d540edc4313e8f0200f9d51f7f86ea0dd0dc
    at web3.js:3143:20
    at web3.js:6347:15
    at web3.js:5081:36
    at <anonymous>:1:1
```
Supposedly, it's possible for me to push into this PR, but I don't know how, so here's the git diff instead: 
```diff

#git diff
diff --git a/internal/ethapi/api.go b/internal/ethapi/api.go
index 30c7d7d..5736b8a 100644
--- a/internal/ethapi/api.go
+++ b/internal/ethapi/api.go
@@ -1003,11 +1003,11 @@ func (s *PublicTransactionPoolAPI) GetRawTransactionByHash(ctx context.Context,
 func (s *PublicTransactionPoolAPI) GetTransactionReceipt(hash common.Hash) (map[string]interface{}, error) {
        tx, blockHash, blockNumber, index := core.GetTransaction(s.b.ChainDb(), hash)
        if tx == nil {
-               return nil, fmt.Errorf("Missing transaction for hash %v", hash)
+               return nil, fmt.Errorf("Missing transaction for hash %v", hash.String())
        }
        receipt, _, _, _ := core.GetReceipt(s.b.ChainDb(), hash) // Old receipts don't have the lookup data available
        if receipt == nil {
-               return nil, fmt.Errorf("Missing receipt for hash %v", hash)
+               return nil, fmt.Errorf("Missing receipt for hash %v", hash.String())
        }
 
        var signer types.Signer = types.FrontierSigner{}

```

**Comment by rhaps107** [comment]
is it ok, @holiman ?

File: internal/ethapi/api.go [hunk 2]
```diff
@@ -1003,9 +1003,12 @@ func (s *PublicTransactionPoolAPI) GetRawTransactionByHash(ctx context.Context,
 func (s *PublicTransactionPoolAPI) GetTransactionReceipt(hash common.Hash) (map[string]interface{}, error) {
 	tx, blockHash, blockNumber, index := core.GetTransaction(s.b.ChainDb(), hash)
 	if tx == nil {
-		return nil, nil
+		return nil, fmt.Errorf("Missing transaction for hash %s", hash.String())
 	}
```
Comment (karalabe) [inline internal/ethapi/api.go]: Error's should be lower cased, so multiple may be composed nicely. Also I don't think it's necessary to add the hash to the error, it's pretty obvious based on the function call parameter what's not being found. I think `unknown transaction` is fine, no need to make it fancier.

File: internal/ethapi/api.go [hunk 3]
```diff
@@ -1003,9 +1003,12 @@ func (s *PublicTransactionPoolAPI) GetRawTransactionByHash(ctx context.Context,
...
 	receipt, _, _, _ := core.GetReceipt(s.b.ChainDb(), hash) // Old receipts don't have the lookup data available
+	if receipt == nil {
+		return nil, fmt.Errorf("Missing receipt for hash %s", hash.String())
+	}
```
Comment (karalabe) [inline internal/ethapi/api.go]: Similarly, I'd just go with `unknown receipt`.

Comment (karalabe) [reply internal/ethapi/api.go]: Although, come to think of it, shouldn't this case be impossible in the first place?

**Comment by GitCop** [comment]
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: caf4c55e5001ce7cc3541d26af2c585184714e9e
  * Commits must be prefixed with the package(s) they modify

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

File: internal/ethapi/api.go (hunk 3 above)
Comment (holiman) [reply internal/ethapi/api.go]: I'd agree, but it obviously isn't, judging from the panics. We should probably look into how that can occur.

Comment (karalabe) [reply internal/ethapi/api.go]: Last time we did such a thing it turned into a very very nasty bug that's still not completely fixed (https://github.com/ethereum/go-ethereum/pull/15235).

Comment (rhaps107) [reply internal/ethapi/api.go]: @karalabe
in first comment you can see mention to issues, when geth panics on getTransactionReceipt.
thats happened to me 3 times.
thats why i created this request. at least node wont shutdown on getTransactionReceipt.
i agree that something strange happened to blocks database in this case.

**Pull request merged**